diff --git a/deluge/core/alertmanager.py b/deluge/core/alertmanager.py
index 24f47fddd..8bc754647 100644
--- a/deluge/core/alertmanager.py
+++ b/deluge/core/alertmanager.py
@@ -67,10 +67,10 @@ def update(self):
         pass
 
     def start(self):
-        thread = threading.Thread(
+        self._thread = threading.Thread(
             target=self.wait_for_alert_in_thread, name='alert-poller', daemon=True
         )
-        thread.start()
+        self._thread.start()
         self._event.set()
 
     def stop(self):
@@ -88,10 +88,13 @@ def wait_for_alert_in_thread(self):
                 time.sleep(0.05)
                 continue
 
+            if not self._event.wait(timeout=1):
+                continue
+
             if self.session.wait_for_alert(1000) is None:
                 continue
-            if self._event.wait():
-                threads.blockingCallFromThread(reactor, self.maybe_handle_alerts)
+
+            threads.blockingCallFromThread(reactor, self.maybe_handle_alerts)
 
     def on_delayed_call_timeout(self, result, timeout, **kwargs):
         log.warning('Alert handler was timed-out before being called %s', kwargs)
